sklearn/utils/estimator_checks.py

```diff
@@ -395,57 +402,145 @@ def _yield_all_checks(estimator, legacy: bool):
+    ----------
+    estimator : estimator object
+        Estimator instance for which to generate checks.
+    check : partial or callable
+        Check to be marked.
+    expected_failed_checks : dict[str, str], default=None
+        Dictionary of the form {check_name: reason} for checks that are expected to
+        fail.
 
-    return wrapped
+    Returns
+    -------
+    should_be_marked : bool
+        Whether the check should be marked as xfail or skipped.
+    reason : str
+        Reason for skipping the check.
+    """
+
+    expected_failed_checks = expected_failed_checks or {}
+
+    check_name = _check_name(check)
+    if check_name in expected_failed_checks:
+        return True, expected_failed_checks[check_name]
 
+    return False, "Check is not expected to fail"
 
-def _should_be_skipped_or_marked(estimator, check):
-    # Return whether a check should be skipped (when using check_estimator())
-    # or marked as XFAIL (when using @parametrize_with_checks()), along with a
-    # reason.
-    # Currently, a check should be skipped or marked if
-    # the check is in the _xfail_checks tag of the estimator
 
-    check_name = check.func.__name__ if isinstance(check, partial) else check.__name__
+def checks_generator(
+    estimator,
+    *,
+    legacy: bool = True,
+    expected_failed_checks: dict[str, str] | None = None,
+    mark: Literal["xfail", "skip", None] = None,
+):
+    """Generate checks for an estimator.
+
+    Parameters
+    ----------
+    estimator : estimator object
+        Estimator instance for which to generate checks.
+    legacy : bool, default=True
+        Whether to include legacy checks. Over time we remove checks from this category
+        and move them into their specific category.
+    expected_failed_checks : dict[str, str], default=None
+        Dictionary of the form {check_name: reason} for checks that are expected to
+        fail.
+    mark : "xfail" or "skip" or None, default=None
+        Whether to mark the checks that are expected to fail as
+        xfail(`pytest.mark.xfail`) or skip. Skip will raise a
+        :class:`~sklearn.exceptions.SkipTest` for the test.
 
-    xfail_checks = get_tags(estimator)._xfail_checks or {}
-    if check_name in xfail_checks:
-        return True, xfail_checks[check_name]
+    Returns
+    -------
+    checks_generator : generator
+        Generator that yields (estimator, check) tuples.
+    """
+    if mark == "xfail":
+        import pytest
+    else:
+        pytest = None  # type: ignore
 
-    return False, "placeholder reason that will never be used"
+    name = type(estimator).__name__
+    # First check that the estimator is cloneable which is needed for the rest
+    # of the checks to run
+    yield estimator, partial(check_estimator_cloneable, name)
+    for check in _yield_all_checks(estimator, legacy=legacy):
+        check_with_name = partial(check, name)
+        for check_instance in _yield_instances_for_check(check, estimator):
```

Comment:
Should we find a way to expose custom instance generation to third-party developers? It feels wrong to have `sklearn.utils.estimator_checks` import stuff from `sklearn.utils._test_common.instance_generator`. I would have expected `sklearn.tests.test_common` to be responsible for this dependency injection.

Note that this design issue was already present in `main`. But since we are to introduce a new public method as part of the developer API, maybe this is the right time to fix it? Possibly by making it possible to pass a dict with the structure of entries of `PER_ESTIMATOR_CHECK_PARAMS` similarly to what we do with the `expected_failed_checks` dict.

Reply:
yes, I have exactly this in mind as a separate PR. This one is already quite large.